components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthUtil.java

```diff
@@ -790,6 +824,109 @@ public static boolean revokeTokens(String username, UserStoreManager userStoreMa
             }
         }
 
+        return isErrorOnRevokingTokens;
+    }
+
+    /**
+     * This method will revoke the access tokens of user.
+     * @param username username.
+     * @param userStoreManager userStoreManager.
+     * @param roleId roleId.
+     * @return true if revocation is successful. Else return false.
+     */
+    public static boolean revokeTokens(String username, UserStoreManager userStoreManager, String roleId)
+            throws UserStoreException {
+
+        String userStoreDomain = UserCoreUtil.getDomainName(userStoreManager.getRealmConfiguration());
+        String tenantDomain = IdentityTenantUtil.getTenantDomain(userStoreManager.getTenantId());
+        AuthenticatedUser authenticatedUser = buildAuthenticatedUser(username, userStoreDomain, tenantDomain);
+
+        /* This userStoreDomain variable is used for access token table partitioning. So it is set to null when access
+        token table partitioning is not enabled.*/
+        userStoreDomain = null;
+        if (OAuth2Util.checkAccessTokenPartitioningEnabled() && OAuth2Util.checkUserNameAssertionEnabled()) {
+            try {
+                userStoreDomain = OAuth2Util.getUserStoreForFederatedUser(authenticatedUser);
+            } catch (IdentityOAuth2Exception e) {
+                LOG.error("Error occurred while getting user store domain for User ID : " + authenticatedUser, e);
+                throw new UserStoreException(e);
+            }
+        }
+
+        // Get details about the role to identify the audience and associated applications.
+        Role role;
+        try {
+            RoleManagementService roleV2ManagementService =
+                    OAuthComponentServiceHolder.getInstance().getRoleV2ManagementService();
+            role = roleV2ManagementService.getRole(roleId, tenantDomain);
+        } catch (IdentityRoleManagementException e) {
+            String errorMessage = "Error occurred while retrieving role of id : " + roleId;
+            throw new UserStoreException(errorMessage, e);
+        }
+
+        Set<String> clientIds;
+        if (role != null && RoleConstants.APPLICATION.equals(role.getAudience())) {
+            // Get clientIds of associated applications for the specific application role.
+            clientIds = getClientIdsOfAssociatedApplications(role, tenantDomain);
```

Comment:
You can have this logic executed if role-id present. Finally we need set of client-ids. Let's avoid code duplications as much possible.

Reply:
```    
    public static boolean revokeTokens(String username, UserStoreManager userStoreManager) throws UserStoreException {

        abc(username, userStoreManager, null);
    }

    public static boolean revokeTokens(String username, UserStoreManager userStoreManager, String roleId)
            throws UserStoreException {

        if (roleId == null) {
            return revokeTokens(username, userStoreManager);
        }

        // Get details about the role to identify the audience and associated applications.
        Role role;
        try {
            RoleManagementService roleV2ManagementService =
                    OAuthComponentServiceHolder.getInstance().getRoleV2ManagementService();
            role = roleV2ManagementService.getRole(roleId, tenantDomain);
            if (!RoleConstants.APPLICATION.equals(role.getAudience())) {
                return revokeTokens(username, userStoreManager);
            }
        } catch (IdentityRoleManagementException e) {
            String errorMessage = "Error occurred while retrieving role of id : " + roleId;
            throw new UserStoreException(errorMessage, e);
        }
        // Get clientIds of associated applications for the specific application role.
        Set<String> clientIds = getClientIdsOfAssociatedApplications(role, tenantDomain);
        abc(username, userStoreManager, clientIds);
    }

    // The common logic
    private void abc(String username, UserStoreManager userStoreManager, Set<String> clientIds) {

        String userStoreDomain = UserCoreUtil.getDomainName(userStoreManager.getRealmConfiguration());
        String tenantDomain = IdentityTenantUtil.getTenantDomain(userStoreManager.getTenantId());
        AuthenticatedUser authenticatedUser = buildAuthenticatedUser(username, userStoreDomain, tenantDomain);

        /* This userStoreDomain variable is used for access token table partitioning. So it is set to null when access
        token table partitioning is not enabled.*/
        userStoreDomain = null;
        if (OAuth2Util.checkAccessTokenPartitioningEnabled() && OAuth2Util.checkUserNameAssertionEnabled()) {
            try {
                userStoreDomain = OAuth2Util.getUserStoreForFederatedUser(authenticatedUser);
            } catch (IdentityOAuth2Exception e) {
                LOG.error("Error occurred while getting user store domain for User ID : " + authenticatedUser, e);
                throw new UserStoreException(e);
            }
        }

        if (clientIds == null) {
            try {
                // get all the distinct client Ids authorized by this user
                clientIds = OAuthTokenPersistenceFactory.getInstance()
                        .getTokenManagementDAO().getAllTimeAuthorizedClientIds(authenticatedUser);
            } catch (IdentityOAuth2Exception e) {
                LOG.error("Error occurred while retrieving apps authorized by User ID : " + authenticatedUser, e);
                throw new UserStoreException(e);
            }
        }

        boolean isErrorOnRevokingTokens;
        isErrorOnRevokingTokens = processTokenRevocation(clientIds, authenticatedUser, userStoreDomain, username);

        // Throw exception if there was any error found in revoking tokens.
        if (isErrorOnRevokingTokens) {
            throw new UserStoreException("Error occurred while revoking Access Tokens of the user " + username);
        }
        return true;
    }````